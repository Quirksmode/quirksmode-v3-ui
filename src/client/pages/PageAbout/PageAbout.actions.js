export const FETCH_ABOUT_DATA = 'fetch_about_data';
export const fetchAboutData = () => async (dispatch, getState, api) => {
  const res = await api.get('quirksmode/v1/pages/about-me');

  dispatch({
    type: FETCH_ABOUT_DATA,
    payload: res
  });
};
