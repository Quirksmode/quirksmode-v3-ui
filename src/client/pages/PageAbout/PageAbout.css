.PageAbout {
  &__col {
    @media (--tablet-portrait) {
      width: 64.8%;
    }
  }

  &__title {
    align-items: center;
    color: var(--color-qm-heading);
    display: flex;
    margin-bottom: 30px;
  }

  &__introText {
    color: var(--color-qm-heading);
    font-size: 1.9rem;
    font-weight: 200;
    line-height: 3.3rem;
    margin-bottom: 20px;
    position: relative;
    text-rendering: optimizelegibility;
    z-index: var(--z-index-PageAbout__introText);

    @media (--tablet-portrait) {
      width: 64.8%;
    }
  }

  &__sketch {
    &::before {
      @media (--tablet-portrait) {
        background: url(../img/sketch_about_2x.png) 0 0 no-repeat;
        background-size: 785px 1550px;
        content: '';
        height: 1550px;
        pointer-events: none;
        position: absolute;
        right: -130px;
        top: -446px;
        width: 785px;
        z-index: var(--z-index-PageAbout__sketch__before);
      }
    }
  }

  &__skillsWrap {
    @media (--custom-570) {
      display: flex;
      flex-wrap: wrap;
      justify-content: space-between;
    }

    @media (--tablet-portrait) {
      display: block;
    }

    @media (--custom-1000) {
      display: flex;
      flex-wrap: wrap;
      justify-content: space-between;
    }
  }

  &__skills {
    margin-bottom: 35px;
    width: 100%;

    @media (--custom-570) {
      width: 47.2%;
    }

    @media (--tablet-portrait) {
      width: 80%;
    }

    @media (--custom-1000) {
      width: 47.2%;
    }

    &.first {
      @media (--custom-570) {
        margin-left: 0;
      }
    }

    h3 {
      color: var(--color-emperor);
      font-family: var(--font-body-main);
      font-size: 1.6rem;
      font-weight: 300;
      line-height: 1.6rem;
      margin-bottom: 10px;
    }
  }

  &__skillBox {
    border-radius: 3px;
    margin-bottom: 1px;

    &.grey {
      background-color: var(--color-sb-grey);
    }

    &.blue {
      background-color: var(--color-sb-blue);
    }

    &.red {
      background-color: var(--color-primary-red-washed);
    }

    &.w20 {
      width: 20%;
    }

    &.w40 {
      width: 40%;
    }

    &.w60 {
      width: 60%;
    }

    &.w80 {
      width: 80%;
    }

    &.w100 {
      width: 100%;
    }

    p {
      color: var(--color-white);
      font-size: 1.4rem;
      line-height: 1.4rem;
      margin: 0;
      padding: 9px 0 9px 9px;

      span {
        color: var(--color-qm-middlebody);
        display: inline-block;
        margin-left: 5px;
      }
    }
  }

  &__cvSection {
    ul {
      margin-bottom: 60px;
    }

    li {
      font-size: 1.5rem;
      margin-bottom: 40px;

      h3 {
        color: var(--color-qm-heading);
        font-family: var(--font-body-main);
        font-size: 1.7rem;
        font-weight: 500;
        line-height: 2.2rem;
        margin-bottom: 10px;
      }

      ul {
        list-style-type: square;
        margin: 10px 0 10px 25px;

        li {
          margin-bottom: 3px;

          ul {
            list-style-type: square;
          }
        }

        a {
          display: block;
        }
      }
    }

    span {
      color: var(--color-qm-middlebody);
      display: block;
      font-size: 1.5rem;
      font-weight: 300;
    }

    &--education {
      span {
        color: var(--color-qm-middlebody);
        display: inline-block;
      }

      li {
        margin-bottom: 20px;

        li {
          margin-bottom: 0;
        }
      }
    }

    &--training {
      ul {
        ul {
          @media (--mobile-landscape) {
            column-gap: 60px;
            columns: 2;
          }
        }
      }

      li {
        break-inside: avoid;
        page-break-inside: avoid;
      }

      span {
        display: block;

        @media (--tablet-portrait) {
          display: inline;
        }
      }
    }

    &--interests {
      p {
        font-size: 1.5rem;
      }
    }
  }
}
