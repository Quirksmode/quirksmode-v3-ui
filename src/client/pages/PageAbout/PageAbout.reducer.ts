import {
  FETCH_ABOUT_REQUEST,
  FETCH_ABOUT_SUCCESS,
  FETCH_ABOUT_ERROR
} from './PageAbout.actions';

export const initialState = {
  content: {
    title: '',
    intro: '',
    skillsSections: {
      skillsIcon: '',
      skillsTitle: '',
      skills: []
    },
    cvSections: []
  },
  metadata: {},
  loading: false,
  error: false
};

export default (state = initialState, action) => {
  switch (action.type) {
    case FETCH_ABOUT_REQUEST:
      return {
        ...state,
        loading: true,
        error: false
      };
    case FETCH_ABOUT_SUCCESS:
      return {
        ...state,
        content: action.payload.data.content,
        metadata: action.payload.data.metadata,
        loading: false,
        error: false
      };
    case FETCH_ABOUT_ERROR:
      return {
        ...state,
        loading: false,
        error: true
      };
    default:
      return state;
  }
};
