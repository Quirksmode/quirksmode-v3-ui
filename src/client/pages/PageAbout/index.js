import React from 'react';
import loadable from '@loadable/component';
import Loading from 'components/Loading/Loading';
import ErrorBoundary from 'components/ErrorBoundary/ErrorBoundary';

const PageAbout = loadable(() => import(/* webpackPrefetch: true */ './PageAbout'), {
  fallback: <Loading />
});

const WrappedPageAbout = props => (
  <ErrorBoundary>
    <PageAbout { ...props } />
  </ErrorBoundary>
);

export default WrappedPageAbout;
