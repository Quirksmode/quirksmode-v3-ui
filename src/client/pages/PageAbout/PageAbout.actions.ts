import axios from 'axios';
import { PageAboutData } from './PageAbout.types';
import {
  fetchAboutRequest,
  fetchAboutSuccess,
  fetchAboutError,
} from './PageAbout.reducer';
import { AppThunk } from 'client/redux/types';

export const fetchAboutData = (): AppThunk => async (dispatch) => {
  dispatch(fetchAboutRequest());

  try {
    const res = await axios.get<PageAboutData>(
      `${process.env.CMS_URL}/wp-json/quirksmode/v1/pages/about-me`
    );
    dispatch(fetchAboutSuccess(res.data));
  } catch (err) {
    dispatch(fetchAboutError());
  }
};
