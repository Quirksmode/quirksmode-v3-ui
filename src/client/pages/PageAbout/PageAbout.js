import React, { useEffect } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import SVGInline from 'react-svg-inline';
import Breadcrumbs from 'components/Breadcrumbs/Breadcrumbs';
import Meta from 'components/Meta/Meta';
import PageWrapper from 'components/PageWrapper/PageWrapper';
import IconArrowDown from '!!raw-loader!icons/arrowdown.svg';
import { fetchAboutData } from './PageAbout.actions';

/**
 * About Page
 *
 * @name PageAbout
 * @param {function} props.fetchAboutDataAction [Redux action to Fetch the About Page Data]
 * @param {object} props.content [The content for this Page]
 * @param {object} props.metadata [The metadata for this Page]
 * @param {boolean} props.loading [Flag for while the data is being fetched]
 * @param {boolean} props.error [Flag for if there is an error fetching the data]
 * @param {object} props.siteSettings [The Global Site Settings]
 * @return {JSXElement}
 */
const PageAbout = ({
  fetchAboutDataAction,
  content,
  metadata,
  loading,
  error,
  siteSettings,
}) => {
  /**
   * @type {Object}
   * @property {string} content.title [The Page Title]
   * @property {object} content.intro [Intro content]
   * @property {array} content.skillsSections [The Skills Sections]
   * @property {array} content.cvSections [The CV Sections]
   */
  const {
    title,
    intro,
    skillsSections,
    cvSections
  } = content;

  /**
   * @type {Object}
   * @property {object} siteSettings.cv [CV data from the Global Site Settings]
   */
  const {
    cv
  } = siteSettings;

  /**
   * Fetch the About Page Data via Redux, but only if the data does not already exist
   */
  useEffect(() => {
    if (!title) fetchAboutDataAction();
  }, [fetchAboutDataAction, title]);

  return title && (
    <PageWrapper error={ error } loading={ loading }>
      <div className="Page PageAbout">
        <Meta { ...metadata } />
        <section className="Page__section Page__section--greyFade">
          <div className="Page__sectionInner grid">
            <h1>{ title }</h1>
            <Breadcrumbs>
              <span className="Breadcrumbs__divider">&gt;</span>
              <span className="Breadcrumbs__active">{ title }</span>
            </Breadcrumbs>
            <div className="PageAbout__intro">
              <div
                dangerouslySetInnerHTML={ { __html: intro } }
                className="PageAbout__introText"
              />
              <a
                className="btn--submitIcon"
                href={ cv.url }
              >
                DOWNLOAD MY CV
                <SVGInline svg={ IconArrowDown } />
              </a>
            </div>
          </div>
        </section>
        <section className="Page__section Page__section--greyFade">
          <div className="Page__sectionInner PageAbout__sketch grid">
            <div className="PageAbout__col">
              <h2 className="PageAbout__title">
                { skillsSections.skillsTitle }
              </h2>
              <div className="PageAbout__skillsWrap">
                { skillsSections.skills.map((skillsItem, skillsItemIndex) => {
                  const {
                    add_skill: addSkill,
                    skills_category_name: skillsCategoryName,
                    skills_colour: skillsColour
                  } = skillsItem;

                  return (
                    <div
                      key={ skillsItemIndex }
                      className="PageAbout__skills"
                    >
                      <h3>{ skillsCategoryName }</h3>
                      {
                          addSkill.map((skillItem, skillItemIndex) => {
                            const {
                              skill_name: skillName,
                              skill_percentage: skillPercentage
                            } = skillItem;

                            return (
                              <div
                                key={ skillItemIndex }
                                className={ `PageAbout__skillBox ${skillsColour} w${skillPercentage}` }
                              >
                                <p>
                                  { skillName }
                                </p>
                              </div>
                            );
                          })
                        }
                    </div>
                  );
                })}
              </div>
            </div>
          </div>
        </section>
        <section className="Page__section Page__section--whiteOverlay clearfix">
          <div className="Page__sectionInner grid clearfix">
            <div className="PageAbout__col first clearfix">
              { cvSections.map((cvSection, index) => {
                const {
                  add_class: addClass,
                  add_content: addContent,
                  add_title: addTitle
                } = cvSection;

                return (
                  <div
                    key={ index }
                    className={ `PageAbout__cvSection PageAbout__cvSection--${addClass}` }
                  >
                    <h2 className="PageAbout__title">
                      { addTitle }
                    </h2>
                    <div
                      className="PageAbout__cvSection-content"
                      dangerouslySetInnerHTML={ { __html: addContent } }
                    />
                  </div>
                );
              })}
            </div>
          </div>
        </section>
      </div>
    </PageWrapper>
  );
};

PageAbout.propTypes = {
  fetchAboutDataAction: PropTypes.func,
  content: PropTypes.object,
  metadata: PropTypes.object,
  loading: PropTypes.bool,
  error: PropTypes.bool,
  siteSettings: PropTypes.object,
};

const mapStateToProps = ({ pageAbout, app }) => ({
  content: pageAbout.content,
  metadata: pageAbout.metadata,
  loading: pageAbout.loading,
  error: pageAbout.error,
  siteSettings: app.siteSettings,
});

const mapDispatchToProps = dispatch => ({
  fetchAboutDataAction: (...args) => dispatch(fetchAboutData(...args))
});

export default connect(mapStateToProps, mapDispatchToProps)(PageAbout);
