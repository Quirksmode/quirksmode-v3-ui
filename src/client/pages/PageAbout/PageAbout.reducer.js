import { FETCH_ABOUT_DATA } from './PageAbout.actions';

const initialState = {
  title: '',
  content: '',
  skillsSections: {
    skillsIcon: '',
    skillsTitle: '',
    skills: []
  },
  cvSections: [],
  metadata: {}
};

export default (state = initialState, action) => {
  switch (action.type) {
    case FETCH_ABOUT_DATA:
      return {
        ...state,
        title: action.payload.data.title,
        content: action.payload.data.content,
        skillsSections: action.payload.data.skillsSections,
        cvSections: action.payload.data.cvSections,
        metadata: action.payload.data.metadata
      };
    default:
      return state;
  }
};
