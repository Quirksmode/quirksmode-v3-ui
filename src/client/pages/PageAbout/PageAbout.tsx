import React, { useEffect } from 'react';
import { useDispatch } from 'react-redux';
import Breadcrumbs from 'components/Breadcrumbs/Breadcrumbs';
import Meta from 'components/Meta/Meta';
import PageWrapper from 'components/PageWrapper/PageWrapper';
import IconArrowDown from 'icons/arrowdown.svg';
import { fetchAboutData } from './PageAbout.actions';
import { useTypedSelector } from 'client/redux/types';

/**
 * About Page
 */
const PageAbout: React.FC = () => {
  const dispatch = useDispatch();
  const pageAbout = useTypedSelector((state) => state.pageAbout);
  const app = useTypedSelector((state) => state.app);

  const { content, metadata, loading, error } = pageAbout;
  const { title, intro, skillsSections, cvSections } = content;
  const { cv } = app.siteSettings;

  /**
   * Fetch the About Page Data via Redux, but only if the data does not already exist
   */
  useEffect(() => {
    if (!title) dispatch(fetchAboutData());
  }, [title]);

  return (
    <PageWrapper error={error} loading={loading}>
      <div className="Page PageAbout">
        {metadata && <Meta {...metadata} />}
        <section className="Page__section Page__section--greyFade">
          <div className="Page__sectionInner grid">
            <h1>{title}</h1>
            <Breadcrumbs>
              <span className="Breadcrumbs__divider">&gt;</span>
              <span className="Breadcrumbs__active">{title}</span>
            </Breadcrumbs>
            <div className="PageAbout__intro">
              <div
                dangerouslySetInnerHTML={{ __html: intro }}
                className="PageAbout__introText"
              />
              {cv?.url && (
                <a className="btn--submitIcon" href={cv.url}>
                  DOWNLOAD MY CV
                  <IconArrowDown />
                </a>
              )}
            </div>
          </div>
        </section>
        <section className="Page__section Page__section--greyFade">
          <div className="Page__sectionInner PageAbout__sketch grid">
            <div className="PageAbout__col">
              <h2 className="PageAbout__title">{skillsSections.skillsTitle}</h2>
              <div className="PageAbout__skillsWrap">
                {skillsSections.skills.map((skillsItem, skillsItemIndex) => {
                  const {
                    add_skill: addSkill,
                    skills_category_name: skillsCategoryName,
                    skills_colour: skillsColour,
                  } = skillsItem;

                  return (
                    <div
                      key={skillsItemIndex}
                      className="PageAbout__skills"
                      data-test="PageAbout__skills"
                    >
                      <h3>{skillsCategoryName}</h3>
                      {addSkill.map((skillItem, skillItemIndex) => {
                        const {
                          skill_name: skillName,
                          skill_percentage: skillPercentage,
                        } = skillItem;

                        return (
                          <div
                            key={skillItemIndex}
                            className={`PageAbout__skillBox ${skillsColour} w${skillPercentage}`}
                            data-test="PageAbout__skillBox"
                          >
                            <p>{skillName}</p>
                          </div>
                        );
                      })}
                    </div>
                  );
                })}
              </div>
            </div>
          </div>
        </section>
        <section className="Page__section Page__section--whiteOverlay clearfix">
          <div className="Page__sectionInner grid clearfix">
            <div className="PageAbout__col first clearfix">
              {cvSections.map((cvSection, index) => {
                const {
                  add_class: addClass,
                  add_content: addContent,
                  add_title: addTitle,
                } = cvSection;

                return (
                  <div
                    key={index}
                    className={`PageAbout__cvSection PageAbout__cvSection--${addClass}`}
                    data-test="PageAbout__cvSection"
                  >
                    <h2 className="PageAbout__title">{addTitle}</h2>
                    <div
                      className="PageAbout__cvSection-content"
                      dangerouslySetInnerHTML={{ __html: addContent }}
                    />
                  </div>
                );
              })}
            </div>
          </div>
        </section>
      </div>
    </PageWrapper>
  );
};

export default PageAbout;
