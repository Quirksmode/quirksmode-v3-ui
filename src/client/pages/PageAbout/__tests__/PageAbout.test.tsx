import {
  findByTestAttr,
  mountRender,
  reduxHelper,
} from 'client/tests/testUtils';
import moxios from 'moxios';
import PageAbout from '../PageAbout';
import AppStub from 'client/__tests__/app.stub.json';
import PageAboutStub from './PageAbout.stub.json';
import PageWrapper from 'client/components/PageWrapper/PageWrapper';

describe('PageAbout', () => {
  let wrapper;
  const props = {};
  const state = {
    app: AppStub,
    pageAbout: PageAboutStub,
  };

  beforeEach(() => {
    moxios.install();
  });

  afterEach(() => {
    moxios.uninstall();
  });

  it('matches snapshot and renders correctly with data', () => {
    const { ReduxWrapper } = reduxHelper(PageAbout, props, state);
    wrapper = mountRender(ReduxWrapper);
    expect(wrapper.find(PageWrapper).length).toEqual(1);
    expect(findByTestAttr(wrapper, 'Breadcrumbs').length).toEqual(1);
    expect(findByTestAttr(wrapper, 'PageAbout__skills').length).toBe(8);
    expect(findByTestAttr(wrapper, 'PageAbout__skillBox').length).toBe(40);
    expect(findByTestAttr(wrapper, 'PageAbout__cvSection').length).toBe(4);
    expect(wrapper.html).toMatchSnapshot();
  });

  it('renders correctly when there is no data', () => {
    const { ReduxWrapper } = reduxHelper(PageAbout, props);
    wrapper = mountRender(ReduxWrapper);
    expect(wrapper.find(PageWrapper).length).toEqual(1);
    expect(wrapper.html).toMatchSnapshot();
  });

  it('runs the fetchAboutData action if the title is not defined', () => {
    const newState = {
      ...state,
      pageAbout: {
        ...state.pageAbout,
        content: {
          ...state.pageAbout.content,
          title: '',
        },
      },
    };
    const { dispatch, ReduxWrapper } = reduxHelper(PageAbout, props, newState);
    wrapper = mountRender(ReduxWrapper);
    expect(dispatch).toHaveBeenCalledTimes(1);
  });

  it("doesn't run the fetchAboutData action if the title is defined", () => {
    const { dispatch, ReduxWrapper } = reduxHelper(PageAbout, props, state);
    wrapper = mountRender(ReduxWrapper);
    expect(dispatch).toHaveBeenCalledTimes(0);
  });

  it('runs the correct actions when the About data is successfully fetched', async (done) => {
    const { getActions, ReduxWrapper } = reduxHelper(PageAbout, props);
    wrapper = mountRender(ReduxWrapper);

    moxios.wait(() => {
      const request = moxios.requests.mostRecent();
      request
        .respondWith({
          status: 200,
          response: PageAboutStub,
        })
        .then(() => {
          const expected = [
            {
              type: 'pageAbout/fetchAboutRequest',
            },
            {
              type: 'pageAbout/fetchAboutSuccess',
              payload: PageAboutStub,
            },
          ];
          const actual = getActions();
          expect(actual).toEqual(expected);
          done();
        });
    });
  });

  it('runs the correct actions when the About data is not successfully fetched', async (done) => {
    const { getActions, ReduxWrapper } = reduxHelper(PageAbout, props);
    wrapper = mountRender(ReduxWrapper);

    moxios.wait(() => {
      const request = moxios.requests.mostRecent();
      request
        .respondWith({
          status: 500,
        })
        .then(() => {
          const expected = [
            {
              type: 'pageAbout/fetchAboutRequest',
            },
            {
              type: 'pageAbout/fetchAboutError',
            },
          ];
          const actual = getActions();
          expect(actual).toEqual(expected);
          done();
        });
    });
  });
});
