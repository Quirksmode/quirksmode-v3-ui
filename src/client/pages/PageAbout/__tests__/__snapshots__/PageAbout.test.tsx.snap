// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PageAbout matches snapshot and renders correctly with data 1`] = `[Function]`;

exports[`PageAbout renders correctly when there is no data 1`] = `[Function]`;
